### PR TITLE
Feature/181 generate challenge ai

### PR DESCRIPTION
**Pull request by bodichka11**

### CODE REVIEW

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/AnalyzeCodeController.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿using LeetWars.CodeAnalyzer.Interfaces;
+using LeetWars.Core.Common.DTO.CodeAnalysis;
+using Microsoft.AspNetCore.Mvc;
+
+namespace LeetWars.CodeAnalyzer.WebAPI.Controllers
+{
```
Comment (pospyra) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/AnalyzeCodeController.cs:6]: [Route("[controller]")]
 and delete route from [HttpPost]

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/ChallengeGenerateController.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using LeetWars.CodeAnalyzer.Interfaces;
+using LeetWars.Core.Common.DTO.ChallengeRequest;
+using Microsoft.AspNetCore.Mvc;
+
+namespace LeetWars.CodeAnalyzer.Controllers
+{
+    [ApiController]
```
Comment (pospyra) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/ChallengeGenerateController.cs:7]: the same

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            challengeGenerateResponseDto.Level = challengeGenerateRequest.Level;
+            challengeGenerateResponseDto.Category = challengeGenerateRequest.Category;
+            challengeGenerateResponseDto.Tags = challengeGenerateRequest.Tags;
+            challengeGenerateResponseDto.Language = challengeGenerateRequest.Language;
+            challengeGenerateResponseDto.Description = challengeResponse.Description;
+            challengeGenerateResponseDto.CompleteSolution = challengeResponse.CompleteSolution;
+            challengeGenerateResponseDto.InitialSolution = challengeResponse.InitialSolution;
+            challengeGenerateResponseDto.TestCases = challengeResponse.Tests;
+            challengeGenerateResponseDto.ExampleTestCases = challengeResponse.TestsSubset;
+
```
Comment (pospyra) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs]: maybe mapper?

**Comment by bodichka11** [comment]
https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/113329862/961e04d2-41c6-47a7-aa70-80b880009188

https://trello.com/c/hOkdqz0N/181-generate-challenge-ai

**Review by pospyra — APPROVED** [review]

**Comment by tatianahutii** [comment]
@bodichka11 please rework modal windows to our common style

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/AnalyzeCodeController.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            _analyzeCodeService = analyzeCodeService;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> RunCodeAnalysisAsync(CodeRequestAnalysisDto requestDto)
+        {
+            var response = await _analyzeCodeService.RunCodeAnalysisAsync(requestDto);
+            return Ok(response);
+
+
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Controllers/AnalyzeCodeController.cs]: empty lines

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/MappingProfiles/ChallengeGeneratorProfile.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+                .ForMember(dest => dest.Level, opt => opt.MapFrom(src => src.Level))
+                .ForMember(dest => dest.Tags, opt => opt.MapFrom(src => src.Tags))
+                .ForMember(dest => dest.Language, opt => opt.MapFrom(src => src.Language));
+
+            CreateMap<OpenAiChallengeGenerateResponseDto, ChallengeGenerateResponseDto>()
+                .ForMember(dest => dest.Description, opt => opt.MapFrom(src => src.Description))
+                .ForMember(dest => dest.CompleteSolution, opt => opt.MapFrom(src => src.CompleteSolution))
+                .ForMember(dest => dest.InitialSolution, opt => opt.MapFrom(src => src.InitialSolution))
+                .ForMember(dest => dest.TestCases, opt => opt.MapFrom(src => src.Tests))
+                .ForMember(dest => dest.ExampleTestCases, opt => opt.MapFrom(src => src.TestsSubset));
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/MappingProfiles/ChallengeGeneratorProfile.cs]: you dont need to defind map for all properties with the same name

and you can use .ReverseMap()

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/OpenAISettings/OpenAiSettings.cs [hunk 6]
```diff
@@ -18,5 +19,35 @@ public class OpenAiSettings : IOpenAiSettings
             "Your answer must contain only percent of overall quality mark. (Without any text and '%' symbol). " +
             "Don't provide additional info, only a number." +
             $"Code listing: \n ${codeListing}";
+
+        public string GetChallengePrompt(ChallengeGenerateRequestDto challengeGenerateRequestDto)
+        {
+            if(challengeGenerateRequestDto.Tags == null)
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/OpenAISettings/OpenAiSettings.cs]:
```suggestion
            if (challengeGenerateRequestDto.Tags is null) 
```

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            string prompt = _openAiSettings.GetChallengePrompt(challengeGenerateRequest);
+
+            var response = await GenerateTextResponse(prompt);
+
+            var challengeResponse = JsonConvert.DeserializeObject<OpenAiChallengeGenerateResponseDto>(response);
+
+            if (challengeResponse == null)
+            {
+                throw new NotFoundException(nameof(challengeResponse));
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs]:
```suggestion
            var challengeResponse = JsonConvert.DeserializeObject<OpenAiChallengeGenerateResponseDto>(response) ??  throw new NotFoundException(nameof(challengeResponse));
```

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+            var challengeGenerateResponseDto = _mapper.Map<OpenAiChallengeGenerateResponseDto, ChallengeGenerateResponseDto>(challengeResponse);
+            _mapper.Map(challengeGenerateRequest, challengeGenerateResponseDto);
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/ChallengeGenerator.cs:32]: why do we need 2 maps?

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/CompletionService.cs [hunk 9]
```diff
@@ -20,6 +20,7 @@ public async Task<ChatResult> CreateCompletionAsync(string prompt)
         {
             var request = new ChatRequest()
             {
+
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/Services/CompletionService.cs]: empty line

File: backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/appsettings.json [hunk 10]
```diff
@@ -11,5 +11,6 @@
     "Temperature": "0.5",
     "PresencePenalty": "0.1",
     "FrequencyPenalty": "0.1"
-  }
+  },
+
```
Comment (tatianahutii) [inline backend/LeetWars.CodeAnalyzer/LeetWars.CodeAnalyzer.WebAPI/appsettings.json:15]: empty line

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/ChallengeController.cs [hunk 11]
```diff
@@ -17,7 +17,7 @@ namespace LeetWars.Core.WebAPI.Controllers
     public class ChallengeController : ControllerBase
     {
         private readonly IChallengeService _challengeService;
-
+
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/ChallengeController.cs]: new empty spaces

File: frontend/nginx.conf [hunk 12]
```diff
@@ -113,5 +113,20 @@ http {
 
       proxy_pass         http://builder;
     }
+
+    location /codeanalyzer {
```
Comment (tatianahutii) [inline frontend/nginx.conf:117]: please check upstream as well

File: frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts [hunk 13]
```diff
@@ -84,6 +87,34 @@ export class ChallengeCreationComponent extends BaseComponent implements HasUnsa
...
+            this.challenge.instructions = description;
+            this.challenge.category = category;
+            this.challenge.level = level;
+            this.challenge.tags = tags;
+            this.challengeVersion.completeSolution = completeSolution;
+            this.challengeVersion.initialSolution = initialSolution;
+            this.challengeVersion.exampleTestCases = exampleTestCases;
+            this.challengeVersion.testCases = testCases;
+            this.challengeVersion.languageId = language?.id || 0;
+        }
```
Comment (tatianahutii) [inline frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts]: not in constructore
and use spread operator for defending properties

and why || 0? default values should be null or undefined

File: frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts [hunk 14]
```diff
@@ -107,7 +138,7 @@ export class ChallengeCreationComponent extends BaseComponent implements HasUnsa
         });
     }
 
-    onStepClick(step: ChallengeStep) {
+    public onStepClick(step: ChallengeStep) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts]: we dont need public here

File: frontend/src/app/modules/main/main-page/main-page.component.html [hunk 15]
```diff
@@ -2,9 +2,9 @@
     <div class="container-fluid">
         <div class="new-challenge">
             <button [routerLink]="'/challenges/create'">Create new challenge</button>
+            <button class="generate-challenge" (click)="onBtnGenerateClick()">Generate Challenge by AI</button>
         </div>
         <app-suggested-challenge></app-suggested-challenge>
         <app-filtering-section></app-filtering-section>
     </div>
 </div>
-<app-arrow-up [onScrollEvent]="scrollEventSubject.asObservable()"></app-arrow-up>
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.html]: why did you delete this?

File: frontend/src/app/modules/main/main-page/main-page.component.ts [hunk 16]
```diff
@@ -1,9 +1,78 @@
...
+    }
+
+    public onBtnGenerateClick() {
+        const modalRef = this.modalService.open(DescribeQuestionComponent);
+
+        modalRef.componentInstance.buttons = [
+            {
+                text: 'Generate',
+                handler: () => {
+                    modalRef.componentInstance.updateChallengeGenerateRequest();
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.ts:38]: why this method is in modal?

File: frontend/src/app/modules/main/main-page/main-page.component.ts [hunk 17]
```diff
@@ -1,9 +1,78 @@
...
+            {
+                text: 'Cancel',
+                handler: () => {
+                    modalRef.dismiss();
+                },
+            },
+        ];
+        modalRef.componentInstance.challengeGenerateRequestChange.subscribe((request: IChallengeGenerateRequest) => {
+            this.challengeGenerateRequest = request;
+        });
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.ts:52]: see the previous comment

File: frontend/src/app/modules/main/main-page/main-page.component.ts [hunk 18]
```diff
@@ -1,9 +1,78 @@
...
+        if (!this.challengeGenerateRequest) {
+            return;
+        }
+        this.spinnerService.show();
+        this.codeAnylizerService
+            .generateChallenge(this.challengeGenerateRequest)
+            .pipe(takeUntil(this.unsubscribe$))
+            .subscribe({
+                next: (response: IChallengeGenerateResponse) => {
+                    this.spinnerService.show();
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.ts]: why spinner show 2 times?

File: frontend/src/app/modules/main/main-page/main-page.component.ts [hunk 19]
```diff
@@ -1,9 +1,78 @@
...
+                    this.router.navigate(['/challenges/create'], {
+                        state: {
+                            myComplexObject: response,
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.ts]: not a good name of state data

File: frontend/src/app/modules/main/main-page/main-page.component.ts [hunk 20]
```diff
@@ -1,9 +1,78 @@
...
+                        },
+                    });
+                    this.toastrService.showSuccess('Challenge generated successfully');
+                },
+                error: () => {
+                    this.toastrService.showError('Error generating challenge');
+                },
+            });
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/main-page/main-page.component.ts:78]: where are you hiding the spinner?

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.sass [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+:host
+  font-family: Inter
+  color: $white
+  display: block
+
+.mt-12px
+  margin-top: 12px
+
+.mr-24px
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.sass]: different indent
should be tab

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 22]
```diff
@@ -0,0 +1,193 @@
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: I beliave this should be a part of challenge module, not shared

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+
+    public challengeVersion: INewChallengeVersion;
+
+    public editorOptions = editorOptions;
+
+    @Input() checkValidation = false;
+
+    @Output() validationChange = new EventEmitter<boolean>();
+
+    @Output() challengeGenerateRequestChange = new EventEmitter<IChallengeGenerateRequest>();
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: all inputs together

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+
+    public updateChallengeGenerateRequest() {
+        const selectedLanguageName = this.currentLanguage?.content;
+        const selectedLanguage: ILanguage | undefined = this.languages.find(
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]:
```suggestion
        const selectedLanguage = this.languages.find(
```

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+            (lang) => lang.name === selectedLanguageName,
+        );
+        const challengeGenerateRequest: IChallengeGenerateRequest = {
+            title: this.challenge.title,
+            category: this.challenge.category,
+            level: this.challenge.level,
+            tags: this.challenge.tags,
+            language: selectedLanguage,
+        };
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: spread operator

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+
+    public selectedLevelName = '';
+
+    @Input() buttons: IModalButton[] = [];
+
+    public customInputHeight = '48px';
+
+    public customInputWidth = '100%';
+
+    public CategoryType = CategoryType;
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: properties all together

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    ngOnInit(): void {
+        this.getTags();
+        this.getChallengeLevels();
+        this.getLanguages();
+        this.updateChallengeGenerateRequest();
+    }
+
+    public onLanguageChanged(selectedItem: IDropdownItem) {
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: you dont need public
please delete everywhere

File: frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+        this.currentLanguage = selectedItem;
+        const language = this.languages.find((l) => l.name === selectedItem.content);
+
+        if (!language) {
+            return;
+        }
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/describe-question/describe-question/describe-question.component.ts]: duplications

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**